### DISCUSSION TITLE
How to modify the request body using a hook

### DISCUSSION THREAD

**Discussion by finnbuga** [sindresorhus/ky#436] [Q&A] [answered]
I'm looking to automatically transform from camel case to snake case all objects I'm sending in a POST request. Thought about using the `beforeRequest` hook. But can't figure out how to modify the body or the json object. The `Request` object doesn't seem to have a `body` property set.

**Comment by sholladay** [comment, marked as answer]
[Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) objects are immutable and while they don't have a `body` property, they do have various methods for reading the body.

Ky allows you to return a new request from a `beforeRequest` hook. So something like this should work...

```js
import snakecaseKeys from 'snakecase-keys';
import ky from 'ky';

const api = ky.extend({
    hooks : {
        beforeRequest : [
            async (request, options) => {
                const json = await request.json();
                const snaked = snakecaseKeys(json);
                return new Request(request.url, {
                    ...options,
                    body : snaked
                });
            }
        ]
    }
});
```

You could further improve this hook by having it short circuit if the `Content-Type` header is incompatible with `.json()`, for example.

**Reply by finnbuga** [reply]
Thanks very much @sholladay. I've built onto your solution and came up with:
```ts
  const json = await request.json();
  const body = JSON.stringify(snakifyCase(json));
  return new Request(request, { body });
```
Basically I made sure that `body` is a string and that the new request has the same headers and all the other properties of the initial request.

**Reply by finnbuga** [reply]
Have initially tried to use 
```ts
return new Request(request.url, {...request, body})
```
However, this doesn't work as the spread syntax doesn't spread ANY property (like `request.method` or `request.headers`). I find this very enigmatic (and frustrating). Seems like none of the `Request` properties are enumerable?

**Reply by sholladay** [reply]
Indeed, the properties of a request instance are not enumerable. It's frustrating, for sure. That said, I imagine they designed it like that to encourage people to use the API "correctly".

Instead of this:

```js
new Request(request.url, {...request, body})
```

They want you to do this:

```js
new Request(request, { body })
```
